File: Autonomous Mobile Robots Course/Chapter17/Lesson4/Chapter17_Lesson4.cpp
// Chapter17_Lesson4.cpp
// Exploration Under Limited Battery/Time (Budget-Aware Frontier Exploration)
// Self-contained demo without ROS.
// Build: g++ -O2 -std=c++17 Chapter17_Lesson4.cpp -o demo
#include <bits/stdc++.h>
using namespace std;

struct Budget { double E, T; };
struct RobotParams { double E_step=1.2, T_step=1.0, E_base=0.10, E_reserve=15.0; };
struct SensorParams { int r=3; };

static inline double clampd(double x,double lo,double hi){ return max(lo,min(hi,x)); }

static double entropyBern(double p){
    p = clampd(p, 1e-9, 1.0-1e-9);
    return -(p*log(p) + (1.0-p)*log(1.0-p));
}

struct Grid {
    int W,H;
    vector<double> p; // row-major p_occ
    Grid(int w,int h):W(w),H(h),p(w*h,0.5){}
    bool inb(int x,int y) const { return 0<=x && x<W && 0<=y && y<H; }
    double get(int x,int y) const { return p[y*W+x]; }
    void setv(int x,int y,double v){ p[y*W+x]=clampd(v,0.0,1.0); }
    bool isFree(int x,int y,double thr=0.65) const { return get(x,y) < thr; }
    bool isOcc(int x,int y,double thr=0.65) const { return get(x,y) > thr; }
};

using Cell = pair<int,int>;

static vector<Cell> neigh4(const Grid& g, Cell c){
    int x=c.first, y=c.second;
    vector<Cell> cand={{x+1,y},{x-1,y},{x,y+1},{x,y-1}};
    vector<Cell> out;
    for(auto &q:cand) if(g.inb(q.first,q.second)) out.push_back(q);
    return out;
}

static vector<Cell> astar(const Grid& g, Cell s, Cell goal, double occ_thr=0.65){
    if(!g.inb(s.first,s.second) || !g.inb(goal.first,goal.second)) return {};
    if(!g.isFree(s.first,s.second,occ_thr) || !g.isFree(goal.first,goal.second,occ_thr)) return {};
    auto h = [&](Cell c){ return abs(c.first-goal.first)+abs(c.second-goal.second); };

    unordered_map<long long,double> gscore;
    unordered_map<long long, long long> parent;
    auto key=[&](Cell c)->long long{ return (long long)c.second*g.W + c.first; };

    struct Node{ double f; Cell c; };
    struct Cmp{ bool operator()(const Node& a,const Node& b) const { return a.f>b.f; } };
    priority_queue<Node, vector<Node>, Cmp> pq;
    pq.push({(double)h(s), s});
    gscore[key(s)] = 0.0;

    while(!pq.empty()){
        auto cur = pq.top().c; pq.pop();
        if(cur==goal){
            vector<Cell> path; path.push_back(cur);
            long long k=key(cur);
            while(parent.count(k)){
                k = parent[k];
                path.push_back({(int)(k%g.W), (int)(k/g.W)});
            }
            reverse(path.begin(), path.end());
            return path;
        }
        long long kc=key(cur);
        double gc=gscore[kc];
        for(auto nb: neigh4(g,cur)){
            if(!g.isFree(nb.first, nb.second, occ_thr)) continue;
            long long kn=key(nb);
            double tent = gc + 1.0;
            if(!gscore.count(kn) || tent < gscore[kn]){
                gscore[kn]=tent;
                parent[kn]=kc;
                pq.push({tent + (double)h(nb), nb});
            }
        }
    }
    return {};
}

static vector<Cell> detectFrontiers(const Grid& g, double free_thr=0.35){
    vector<Cell> fr;
    for(int y=0;y<g.H;y++){
        for(int x=0;x<g.W;x++){
            double p=g.get(x,y);
            if(p >= free_thr) continue;
            for(auto nb: neigh4(g,{x,y})){
                double pn=g.get(nb.first, nb.second);
                if(fabs(pn-0.5) < 0.15){ fr.push_back({x,y}); break; }
            }
        }
    }
    return fr;
}

static double expectedIG(const Grid& belief, Cell pose, const SensorParams& sp){
    double ig=0.0;
    for(int dy=-sp.r; dy<=sp.r; dy++){
        for(int dx=-sp.r; dx<=sp.r; dx++){
            int x=pose.first+dx, y=pose.second+dy;
            if(!belief.inb(x,y)) continue;
            double p0 = belief.get(x,y);
            double H0 = entropyBern(p0);
            // simple proxy posterior entropy
            double Hpost = 0.5*entropyBern(0.95) + 0.5*entropyBern(0.05);
            double w = clampd(1.0 - fabs(p0-0.5)*2.0, 0.0, 1.0);
            ig += w * max(0.0, H0 - Hpost);
        }
    }
    return ig;
}

static void applyObs(Grid& belief, const Grid& truth, Cell pose, const SensorParams& sp){
    for(int dy=-sp.r; dy<=sp.r; dy++){
        for(int dx=-sp.r; dx<=sp.r; dx++){
            int x=pose.first+dx, y=pose.second+dy;
            if(!belief.inb(x,y)) continue;
            bool isOcc = truth.isOcc(x,y,0.5);
            belief.setv(x,y, isOcc ? 0.95 : 0.05);
        }
    }
}

static bool pickGoalBudgeted(
    const Grid& belief,
    const vector<Cell>& frontiers,
    Cell cur, Cell home,
    const Budget& budget,
    const RobotParams& rp,
    const SensorParams& sp,
    double lam,
    Cell& bestGoal,
    vector<Cell>& bestPath
){
    double bestU = -1e18;
    bool found=false;
    // subsample if huge
    vector<Cell> cand=frontiers;
    if((int)cand.size()>80){
        std::mt19937 rng(7);
        shuffle(cand.begin(), cand.end(), rng);
        cand.resize(80);
    }

    for(auto g: cand){
        auto path = astar(belief, cur, g, 0.65);
        if(path.empty()) continue;
        int steps = max(0, (int)path.size()-1);
        double Ego = steps*(rp.E_step + rp.E_base);
        double Tgo = steps*rp.T_step;

        auto back = astar(belief, g, home, 0.65);
        if(back.empty()) continue;
        int bsteps = max(0, (int)back.size()-1);
        double Eback = bsteps*(rp.E_step + rp.E_base);
        double Tback = bsteps*rp.T_step;

        if(budget.E - (Ego+Eback) < rp.E_reserve) continue;
        if(budget.T - (Tgo+Tback) < 0) continue;

        double ig = expectedIG(belief, g, sp);
        double U = ig - lam*steps;
        if(U > bestU){
            bestU=U; bestGoal=g; bestPath=path; found=true;
        }
    }
    return found;
}

int main(){
    int W=35, H=25;
    Grid belief(W,H), truth(W,H);

    // truth: mostly free
    for(int y=0;y<H;y++) for(int x=0;x<W;x++) truth.setv(x,y,0.05);

    // borders
    for(int x=0;x<W;x++){ truth.setv(x,0,0.95); truth.setv(x,H-1,0.95); }
    for(int y=0;y<H;y++){ truth.setv(0,y,0.95); truth.setv(W-1,y,0.95); }

    // random rectangles
    std::mt19937 rng(3);
    auto rint=[&](int a,int b){ std::uniform_int_distribution<int> d(a,b); return d(rng); };
    for(int k=0;k<8;k++){
        int x0=rint(3,W-10), y0=rint(3,H-8);
        int ww=rint(3,7), hh=rint(2,5);
        for(int y=y0;y<min(H-1,y0+hh);y++)
            for(int x=x0;x<min(W-1,x0+ww);x++)
                truth.setv(x,y,0.95);
    }

    Cell home={2,2}, cur=home;

    RobotParams rp;
    SensorParams sp;
    Budget bud{160.0, 140.0};

    applyObs(belief, truth, cur, sp);

    int stepsDone=0, goals=0;

    while(true){
        auto fr = detectFrontiers(belief);
        if(fr.empty()){
            cout << "No frontiers left.\n"; break;
        }
        Cell goal; vector<Cell> path;
        bool ok = pickGoalBudgeted(belief, fr, cur, home, bud, rp, sp, 0.35, goal, path);
        if(!ok){
            cout << "No feasible frontier under budget. Returning.\n";
            break;
        }
        for(size_t i=1;i<path.size();i++){
            bud.E -= (rp.E_step + rp.E_base);
            bud.T -= rp.T_step;
            stepsDone++;
            cur = path[i];
            applyObs(belief, truth, cur, sp);
            if(bud.E < rp.E_reserve || bud.T <= 0) break;
        }
        goals++;
        if(bud.E < rp.E_reserve || bud.T <= 0) break;
        if(goals >= 40) { cout << "Stop: demo goal limit.\n"; break; }
    }

    auto back = astar(belief, cur, home, 0.65);
    if(!back.empty()){
        for(size_t i=1;i<back.size();i++){
            bud.E -= (rp.E_step + rp.E_base);
            bud.T -= rp.T_step;
            cur = back[i];
        }
    }
    cout << "Final pose: ("<<cur.first<<","<<cur.second<<") steps="<<stepsDone<<" goals="<<goals<<"\n";
    cout << "Remaining budget: E="<<bud.E<<" T="<<bud.T<<"\n";
    return 0;
}

File: Autonomous Mobile Robots Course/Chapter17/Lesson4/Chapter17_Lesson4.m
% Chapter17_Lesson4.m
% Exploration Under Limited Battery/Time (Budget-Aware Frontier Exploration)
% Self-contained MATLAB demo (no toolboxes required).
%
% Notes for Simulink:
% - You can implement the "budget update" and "goal selection" blocks in Simulink
%   using MATLAB Function blocks, then connect a discrete-time integrator for
%   time and energy and a state machine (Stateflow) for Explore/Return states.

clear; clc; rng(7);

W = 35; H = 25;
belief = 0.5 * ones(H, W);        % p_occ belief
truth  = 0.05 * ones(H, W);       % synthetic truth (mostly free)

% Borders as obstacles
truth(1,:) = 0.95; truth(end,:) = 0.95;
truth(:,1) = 0.95; truth(:,end) = 0.95;

% Random rectangles
rng(3);
for k=1:8
    x0 = randi([4, W-10]); y0 = randi([4, H-8]);
    ww = randi([3, 7]);    hh = randi([2, 5]);
    x1 = min(W-1, x0+ww);  y1 = min(H-1, y0+hh);
    truth(y0:y1, x0:x1) = 0.95;
end

home = [3, 3];   % [x,y] with 1-based indexing
cur  = home;

% Robot/budget params
E_step = 1.2; E_base = 0.10; T_step = 1.0;
E_reserve = 15.0;
E = 160.0; T = 140.0;

% Sensor params
r = 3;
p_occ_obs = 0.95; p_free_obs = 0.05;

% Initial observation
belief = apply_observation(belief, truth, cur, r, p_occ_obs, p_free_obs);

goals = 0; stepsDone = 0;

while true
    frontiers = detect_frontiers(belief, 0.35);
    if isempty(frontiers)
        disp("No frontiers left."); break;
    end

    [goal, path] = pick_goal_budgeted(belief, frontiers, cur, home, ...
        E, T, E_step, E_base, T_step, E_reserve, r, 0.35);

    if isempty(goal)
        disp("No feasible frontier under budget. Returning."); break;
    end

    % Execute path (skip first cell)
    for i = 2:size(path,1)
        E = E - (E_step + E_base);
        T = T - T_step;
        stepsDone = stepsDone + 1;
        cur = path(i,:);
        belief = apply_observation(belief, truth, cur, r, p_occ_obs, p_free_obs);
        if (E < E_reserve) || (T <= 0)
            disp("Budget exhausted mid-path."); break;
        end
    end

    goals = goals + 1;
    if (E < E_reserve) || (T <= 0), break; end
    if goals >= 40, disp("Stop: demo goal limit."); break; end
end

fprintf("Final pose: (%d,%d) steps=%d goals=%d\n", cur(1), cur(2), stepsDone, goals);
fprintf("Remaining budget: E=%.2f T=%.2f\n", E, T);

% --------------------------- Helper functions ---------------------------

function frontiers = detect_frontiers(belief, free_thr)
    [H,W] = size(belief);
    frontiers = [];
    for y=1:H
        for x=1:W
            if belief(y,x) >= free_thr, continue; end
            nbs = neighbors4([x,y], W, H);
            for k=1:size(nbs,1)
                xn=nbs(k,1); yn=nbs(k,2);
                if abs(belief(yn,xn) - 0.5) < 0.15
                    frontiers = [frontiers; x,y]; %#ok<AGROW>
                    break;
                end
            end
        end
    end
end

function nbs = neighbors4(c, W, H)
    x=c(1); y=c(2);
    cand = [x+1,y; x-1,y; x,y+1; x,y-1];
    keep = cand(:,1)>=1 & cand(:,1)<=W & cand(:,2)>=1 & cand(:,2)<=H;
    nbs = cand(keep,:);
end

function belief = apply_observation(belief, truth, pose, r, p_occ_obs, p_free_obs)
    [H,W] = size(belief);
    x0=pose(1); y0=pose(2);
    for dy=-r:r
        for dx=-r:r
            x=x0+dx; y=y0+dy;
            if x<1 || x>W || y<1 || y>H, continue; end
            isOcc = truth(y,x) > 0.5;
            belief(y,x) = (isOcc)*p_occ_obs + (~isOcc)*p_free_obs;
        end
    end
end

function ig = expected_info_gain(belief, pose, r)
    % nats, approximate entropy reduction proxy
    x0=pose(1); y0=pose(2);
    [H,W] = size(belief);
    ig = 0.0;
    Hpost = 0.5*bern_entropy(0.95) + 0.5*bern_entropy(0.05);
    for dy=-r:r
        for dx=-r:r
            x=x0+dx; y=y0+dy;
            if x<1 || x>W || y<1 || y>H, continue; end
            p0 = belief(y,x);
            H0 = bern_entropy(p0);
            w = max(0.0, min(1.0, 1.0 - abs(p0-0.5)*2.0));
            ig = ig + w * max(0.0, H0 - Hpost);
        end
    end
end

function H = bern_entropy(p)
    p = max(1e-9, min(1-1e-9, p));
    H = -(p*log(p) + (1-p)*log(1-p));
end

function [goal, bestPath] = pick_goal_budgeted(belief, frontiers, cur, home, ...
        E, T, E_step, E_base, T_step, E_reserve, r, lam)

    goal = []; bestPath = [];
    bestU = -1e18;

    % subsample
    if size(frontiers,1) > 80
        idx = randperm(size(frontiers,1), 80);
        frontiers = frontiers(idx,:);
    end

    for k=1:size(frontiers,1)
        g = frontiers(k,:);
        path = astar_grid(belief, cur, g, 0.65);
        if isempty(path), continue; end

        steps = size(path,1) - 1;
        Ego = steps*(E_step + E_base);
        Tgo = steps*T_step;

        back = astar_grid(belief, g, home, 0.65);
        if isempty(back), continue; end
        bsteps = size(back,1) - 1;
        Eback = bsteps*(E_step + E_base);
        Tback = bsteps*T_step;

        if (E - (Ego+Eback) < E_reserve), continue; end
        if (T - (Tgo+Tback) < 0), continue; end

        ig = expected_info_gain(belief, g, r);
        U = ig - lam*steps;

        if U > bestU
            bestU = U;
            goal = g;
            bestPath = path;
        end
    end
end

function path = astar_grid(belief, s, g, occThr)
    % 4-neighbor A*; blocks if belief(y,x) > occThr
    [H,W] = size(belief);
    if any(s<1) || s(1)>W || s(2)>H || any(g<1) || g(1)>W || g(2)>H
        path = []; return;
    end
    if belief(s(2),s(1)) > occThr || belief(g(2),g(1)) > occThr
        path = []; return;
    end

    % maps (use linear indices)
    idx = @(x,y) (y-1)*W + x;
    pos = @(k) [mod(k-1,W)+1, floor((k-1)/W)+1];

    open = java.util.PriorityQueue();
    gscore = containers.Map('KeyType','int32','ValueType','double');
    parent = containers.Map('KeyType','int32','ValueType','int32');

    hs = abs(s(1)-g(1)) + abs(s(2)-g(2));
    open.add({hs, idx(s(1),s(2))});
    gscore(idx(s(1),s(2))) = 0.0;

    visited = containers.Map('KeyType','int32','ValueType','logical');

    while ~open.isEmpty()
        node = open.remove();
        curk = node{2};
        if isKey(visited, curk), continue; end
        visited(curk) = true;

        c = pos(curk);
        if all(c==g)
            % reconstruct
            path = c;
            while isKey(parent, curk)
                curk = parent(curk);
                path = [pos(curk); path]; %#ok<AGROW>
            end
            return;
        end

        nbs = neighbors4(c, W, H);
        for i=1:size(nbs,1)
            nb = nbs(i,:);
            if belief(nb(2),nb(1)) > occThr, continue; end
            nbk = idx(nb(1),nb(2));
            tent = gscore(curk) + 1.0;
            if ~isKey(gscore, nbk) || tent < gscore(nbk)
                gscore(nbk) = tent;
                parent(nbk) = curk;
                f = tent + abs(nb(1)-g(1)) + abs(nb(2)-g(2));
                open.add({f, nbk});
            end
        end
    end
    path = [];
end
